### ISSUE TITLE
Memory leak in Deferred or IO.race

### ISSUE THREAD

**Issue by mpilquist** [typelevel/cats-effect#1136]
Consider:

```scala
Deferred[IO, Int].flatMap { d =>
  def loop: IO[Unit] = IO.race(d.get, IO.cede >> IO.unit) >> loop
  loop
}
```

This leaks memory -- each call to `d.get` registers a callback in the internal `LongMap` and they never get removed as a result of losing the race. Bug may be in `IO.race` instead of `Deferred` -- perhaps not calling the unregistration action?

<img width="1048" alt="image" src="https://user-images.githubusercontent.com/118078/91236976-63994d80-e707-11ea-9d9f-718f3eae5337.png">

Result of running above through FS2's leak testing infrastructure:

```
Running leak test deferred
Heap:    4,464,560 total,      +42,296 since start,      +42,296 in last 1 second
Heap:    4,596,984 total,     +174,720 since start,     +132,424 in last 1 second
Heap:    4,679,464 total,     +257,200 since start,      +82,480 in last 1 second
Heap:    4,803,848 total,     +381,584 since start,     +124,384 in last 1 second
Heap:    4,880,024 total,     +457,760 since start,      +76,176 in last 1 second
Heap:    4,959,328 total,     +537,064 since start,      +79,304 in last 1 second
Heap:    5,032,096 total,     +609,832 since start,      +72,768 in last 1 second
Heap:    5,190,248 total,     +767,984 since start,     +158,152 in last 1 second
Heap:    5,264,224 total,     +841,960 since start,      +73,976 in last 1 second
Heap:    5,423,352 total,   +1,001,088 since start,     +159,128 in last 1 second
Heap:    5,542,248 total,   +1,119,984 since start,     +118,896 in last 1 second
Heap:    5,648,088 total,   +1,225,824 since start,     +105,840 in last 1 second
==> X fs2.MemoryLeakSpec.deferred  15.408s munit.FailException: Development/oss/fs2/core/jvm/src/it/scala/fs2/MemoryLeakSpec.scala:63 leak detected - heap dump: /var/folders/58/rc_15mds5vx9119_96_01wsc0000gp/T/fs2-leak-test-7474361769001304354.hprof
62:        case Right(Left(path)) =>
63:          fail(s"leak detected - heap dump: $path")
64:      }.unsafeRunSync()
    at munit.Assertions.fail(Assertions.scala:189)
    at munit.Assertions.fail$(Assertions.scala:14)
    at munit.FunSuite.fail(FunSuite.scala:11)5s
    at fs2.MemoryLeakSpec.leakTestF$$anonfun$2$$anonfun$1(MemoryLeakSpec.scala:63)
```

**Comment by RaasAhsan** [comment]
The issue here is that the current `IO` can cancel between the time the async action is started and the time the finalizer is registered: https://github.com/typelevel/cats-effect/blob/series/3.x/core/shared/src/main/scala/cats/effect/IOFiber.scala#L413 . The fix will be to define `async` as a derived combinator of the more primitive `promise`, which is tracked in #1104.

Gitter discussion: https://gitter.im/typelevel/cats-effect-dev?at=5f459972ec534f584fbfef02

**Comment by djspiewak** [comment]
~~To be clear, we believe this to be the same issue as #1132, and we believe that the most reasonable solution to the problem involves implementing #1104, which provides a natural and composable answer to these kinds of issues.~~

To be clear, the above is nonsense.

![image](https://user-images.githubusercontent.com/752/91353321-54b6a780-e7a8-11ea-9995-2435537594ad.png)

**Comment by RaasAhsan** [comment]
Are they the same issue? The issue in #1132 is that we're aggressively masking `racePair` whereas the issue here is that the registration effect can be canceled before the async finalizer is registered.

**Comment by djspiewak** [comment]
Derp! This is what happens when you comment on issues based on memory without re-reading. Nice catch.

**Comment by RaasAhsan** [comment]
I believe this should be fixed as part of `cont` in 3.0.0-M1; we should never be able to leak a finalizer once its corresponding async action has started. @mpilquist can you check this again?

**Comment by mpilquist** [comment]
Confirmed this is fixed on 3.0.0-M1
